### PR TITLE
Auto Scrolling feature added issue #15 

### LINKED ISSUE #15: Add an option to auto-scroll website
### Subject of the issue
Add an option that helps auto-scroll the website. You can add 3 to 5 speeds for auto scroll

### PR DESCRIPTION
**Pull request by ManavLohia945**
Closes #15 
Hey Diganta! I have added the auto-scrolling feature as discussed by us .I have given three options for speed: slow, medium and fast and the user has to press the enter button to start and press it again to stop.. And to change speed he has to select the speed he wishes and press enter to start..
image

### CODE REVIEW

**Comment by developer-diganta** [comment]
Can you please check the code once more. There is an error in background.js line 151. Also please add the feature at the end.

**Comment by ManavLohia945** [comment]
Ok will check for the error. But I have an opinion on where to place to the feature. If we place it at the end then there is use of auto-scrolling according to me. Instead if the user sees it at the top he/she can use it accordingly.

**Comment by developer-diganta** [comment]
I agree. But I think that things like font size, colors etc are more sort of important here since you know scrolling can also be done using keyboard with a direct button press. I think people would first want to get the visuals corrected and then focus on the functionality. What do you think?

**Comment by ManavLohia945** [comment]
I am sorry But I did not find any error at line 151. From that line onwards the code for new feature starts

**Comment by ManavLohia945** [comment]
Ok np. instead of keeping it at the end we can add this feature below the font style feature or maybe below the remove images section

File: background.js [hunk 1]
```diff
@@ -75,6 +75,80 @@ for (let i = 0; i < links.length; i++) {
...
+document.body.addEventListener('keypress', function (event)
+// autoScrollFeature.addEventListener('click', function()
+{
+    scroling(event);
+}
+, true);
+
+///////////////////////////////////////////////////////////////////////////////
+const imageReader = document.getElementsByClassName('img-read');
+=======
```
Comment (developer-diganta) [inline background.js:151]: I guess this one is causing error. Can you please check once

Comment (ManavLohia945) [reply background.js:151]: Ok will check

**Comment by ManavLohia945** [comment]
If you could tell me where to place this auto scrolling feature and also I will have to do a pull req from the upstream branch to keep my branch updated  so I will have to  commit only once after doing all of this

**Comment by developer-diganta** [comment]
Keep it at the end for this PR

**Comment by ManavLohia945** [comment]
Done !!

**Review by developer-diganta — CHANGES_REQUESTED** [review]
@ManavLohia945 Thanks for the changes! But still it does not seem to be working. I can take a guess as to what can be changed. I think the scrolling action is being done in background.js but I think it should be done in content.js since that is the one we are injecting into the main webpage. The control for scroll should be in background.js but the action should be in content.js. Can you check once. P.S. First do try resolving without all these hassles. If it does not happen, then this may be worth the shot. Do ask doubts if you have any.

**Comment by ManavLohia945** [comment]
I don't understand why is it not working, when I am running using the local path in my pc it works fine for me. Give me some time I will check again and see what is going wrong.

**Review by developer-diganta — CHANGES_REQUESTED** [review]
Hi @ManavLohia945. I think why it is still not working is because you have not used the scroll action in content.js. The scrolling action can only be controlled via control.js while the request is made from background.js. Can you please see that once. If you have any doubt, do not hesitate to ask me or any of the mentors. Also please check the popup.html files once, change it after making a new pull.

**Comment by ManavLohia945** [comment]
I think now it is working. Plz check

**Comment by developer-diganta** [comment]
dino
I tried it again but still it is not working

**Comment by ManavLohia945** [comment]
Sir you will have to press enter to start and again press enter to stop and scroll to the top of the page and then press enter

**Comment by ManavLohia945** [comment]
When you open the extension press enter to start the auto-scrolling and to change speed first select any speed and then scroll to the top and press enter to start and again press enter to stop

**Comment by ManavLohia945** [comment]
You will have to press the enter button after selecting any of the speed and manually scroll upwards first  becoz this feature is at the end .

**Comment by developer-diganta** [comment]
Hi @ManavLohia945 ! I guess you may have misunderstood the issue. The auto scrolling has to be applied to the website not on the extension. So what you can do it use the auto scrolling part in background.js and move it to content.js and then start that scrolling from the extension. If you need any help, just let me or the mentors know.

**Comment by ManavLohia945** [comment]
Ya I guess I will need your help here..

**Comment by developer-diganta** [comment]
Sure. 
@ManavLohia945 So do it like this. You can remove the enter key option and instead add a button on the extension to stop scrolling. Upon clicking this button, send the control to content.js using ```chrome.tabs.query...```. After that you can add an id to the body tag and use that  id to add scrolling action just like you did for the extension. If you have an query, do ask.

**Comment by ManavLohia945** [comment]
@developer-diganta If you could help me we can complete this issue by today. So that I can submit my project on devfolio website as soon as possible

**Comment by harshit-paneri** [comment]
> Yeah sure @ManavLohia945 . @siddhi-244 @harshit-paneri @Vanshiii203 Can anyone help Manav with the issue? You may feel free to ask me if you require anything.

I understand the issue, 
Issues is to work on autoscrolling on the web page 

and @ManavLohia945 did autoscrolling on the extension.

Am I correct @developer-diganta ?

**Comment by developer-diganta** [comment]
> > Yeah sure @ManavLohia945 . @siddhi-244 @harshit-paneri @Vanshiii203 Can anyone help Manav with the issue? You may feel free to ask me if you require anything.
> 
> I understand the issue, Issues is to work on autoscrolling on the web page
> 
> and @ManavLohia945 did autoscrolling on the extension.
> 
> Am I correct @developer-diganta ?

Yup! The extension should have an option that autoscrolls the webpage (ex wikipedia) and not the extension.

**Comment by ManavLohia945** [comment]
So @harshit-paneri, the auto scrolling feature is working for the extension, but we want it to work on the website and not on the extension. So some changes are required for the same. Share your discord id let's connect there and solve it together.

**Comment by developer-diganta** [comment]
`content.js`
```
else if (action == 'auto-scroll') {
        startScroll(request.interval)
  }
      
function startScroll(interval) {
    let id = setInterval(function(event) {
        window.scrollBy(0, 2);
        if ((window.innerHeight + window.scrollY) >= document.body.offsetHeight) {
            stopScroll();
        }
    }, interval);
  id();
    return id;
}

function stopScroll() {
    clearInterval(scrollerID);
}

```

`background.js`
```

slow_speed.addEventListener('click', function (e) {
      chrome.tabs.query({ active: true, currentWindow: true }, function (tabs) {
      chrome.tabs.sendMessage(tabs[0].id, { action: "auto-scroll",interval:15 });
    });
})

```

**Comment by ManavLohia945** [comment]
Thanks a lot for your help.. Could you please check now. I hope it is working fine.

**Review by developer-diganta — CHANGES_REQUESTED** [review]
Hi @ManavLohia945 ! Yup this was the change. Just one last thing. Remove the comments and then it should be good to merge

**Comment by ManavLohia945** [comment]
Done @developer-diganta!! Thanks a lot for helping me out

**Review by developer-diganta — APPROVED** [review]

**Comment by developer-diganta** [comment]
Great work @ManavLohia945 ! Merged!

**Comment by ManavLohia945** [comment]
It was my pleasure working on this issue and knowing you!! Looking forward to contribute more.

**Comment by ManavLohia945** [comment]
If you are okay with it, we can create a new issue to enable the stop scrolling feature. I think that would be really helpful

**Pull request merged**